testing/test_gateway.py

```diff
@@ -645,3 +645,45 @@ def test_main_thread_only_concurrent_remote_exec_deadlock(
             ch.close()
         gw.exit()
         gw.join()
+
+
+def test_concurrent_rinfo(execmodel: gateway_base.ExecModel) -> None:
+    # Use a temporary Group to force the desired execmodel, since
+    # groups used by fixtures tend to use the thread execmodel.
+    group = execnet.Group(execmodel=execmodel.backend)
+    gw = group.makegateway(f"execmodel={execmodel.backend}//popen")
+
+    # For main_thread_only _rinfo will deadlock unless the gateway
+    # has cached the result earlier, but only if sleep_time
+    # exceeds the 1 second grace period in the WorkerGateway
+    # _local_schedulexec method. For other execmodels, a shorter
+    # sleep_time is sufficient to test concurrent _rinfo.
+    sleep_time = 1.5 if execmodel.backend == "main_thread_only" else 0.5
+    try:
+        if execmodel.backend == "main_thread_only":
+            # Cache rinfo like pytest-dist does for backward compatibility,
+            # in order to prevent a deadlock error when pytest-cov requests
+            # rinfo while the main thread is busy with the pytest-xdist
+            # WorkerController's remote_exec call.
+            gw._rinfo()
+            assert hasattr(gw, "_cache_rinfo")
```

Comment:
> @zmedico I admit I'm a bit confused what the test is testing. What is the failure it is testing against?

If we omit the above `gw._rinfo()` call then this test case reproduces the underlying issue from https://github.com/pytest-dev/pytest-xdist/issues/1071:
```python
  File "execnet/testing/test_gateway.py", line 682, in test_concurrent_rinfo
    rinfo = gw._rinfo()
            ^^^^^^^^^^^
  File "execnet/src/execnet/gateway.py", line 87, in _rinfo
    self._cache_rinfo = RInfo(ch.receive())
                              ^^^^^^^^^^^^
  File "execnet/src/execnet/gateway_base.py", line 934, in receive
    raise self._getremoteerror() or EOFError()
execnet.gateway_base.RemoteError: concurrent remote_exec would cause deadlock for main_thread_only execmodel
```

Reply:
Right but that failure is expected right? And we're testing that adding the extra `_rinfo` call prevents it from happening?

Reply:
I'd like to provide some test coverage for the scenario that led to https://github.com/pytest-dev/pytest-xdist/issues/1071, since this sort of coverage could have prevented a broken release of pytest-xdist.